services/config.go

```diff
@@ -0,0 +1,264 @@
+	cs.clientInfo.ProToken = conf.ProToken
+	cs.clientInfo.Country = conf.Country
+	cs.clientInfo.Ip = conf.Ip
+}
+
+func (cs *configService) Stop() {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	if !cs.running {
+		return
+	}
+
+	close(cs.done)
+	cs.running = false
+}
+
+// fetchConfig fetches the current config from the server and updates the client's config if a change
+// has occurred. It returns the extra sleep time received from the server response and any error that
+// occurred.
+func (cs *configService) fetchConfig() (int64, error) {
+	op := ops.Begin("Fetching config")
+	defer op.End()
+
+	newConf, sleep, err := cs.fetch()
+	if err != nil {
+		return 0, op.FailIf(err)
+	}
+
+	cs.lastFetched = time.Now()
+
+	cs.logger.Debug("Received config")
+	curConf := cs.configHandler.GetConfig()
+	if curConf != nil && !configIsNew(curConf, newConf) {
+		op.Set("config_changed", false)
+		cs.logger.Debug("Config is unchanged")
+		return sleep, nil
+	}
+
+	op.Set("config_changed", true)
+
+	cs.updateClientInfo(newConf)
+	cs.configHandler.SetConfig(newConf)
+
+	return sleep, nil
+}
+
+func (cs *configService) fetch() (*apipb.ConfigResponse, int64, error) {
+	confReq := cs.newRequest()
+	buf, err := proto.Marshal(confReq)
+	if err != nil {
+		return nil, 0, fmt.Errorf("unable to marshal config request: %w", err)
+	}
+
+	resp, sleep, err := post(
+		cs.opts.OriginURL,
+		bytes.NewReader(buf),
+		cs.opts.RoundTripper,
+		cs.opts.UserConfig,
+		cs.logger,
+	)
+	if err != nil {
+		return nil, 0, fmt.Errorf("config request failed: %w", err)
+	}
+	defer resp.Close()
+
+	gzReader, err := gzip.NewReader(resp)
+	if err != nil {
+		return nil, sleep, fmt.Errorf("unable to open gzip reader: %w", err)
+	}
+
+	configBytes, err := io.ReadAll(gzReader)
+	gzReader.Close()
+
+	newConf := &apipb.ConfigResponse{}
+	if err = proto.Unmarshal(configBytes, newConf); err != nil {
+		return nil, 0, fmt.Errorf("unable to unmarshal config: %w", err)
+	}
+
+	return newConf, sleep, err
+}
+
+// newRequest returns a new ConfigRequest with the current client info, proxy ids, and the last
+// time the config
+func (cs *configService) newRequest() *apipb.ConfigRequest {
+	conf := cs.configHandler.GetConfig()
+	proxies := []*apipb.ProxyConnectConfig{}
+	if conf != nil { // not the first request
+		proxies = conf.GetProxy().GetProxies()
+	}
+
+	ids := make([]string, len(proxies))
+	for _, proxy := range proxies {
+		ids = append(ids, proxy.GetTrack())
+	}
+
+	confReq := &apipb.ConfigRequest{
+		ClientInfo: cs.clientInfo,
+		Proxy: &apipb.ConfigRequest_Proxy{
+			Ids:         ids,
+			LastRequest: timestamppb.New(cs.lastFetched),
+		},
+	}
+
+	return confReq
+}
+
+// configIsNew returns true if country, proToken, or ip in currInfo differ from new or if new has
+// proxy configs.
+func configIsNew(cur, new *apipb.ConfigResponse) bool {
+	return cur.GetCountry() != new.GetCountry() ||
+		cur.GetProToken() != new.GetProToken() ||
+		len(new.GetProxy().GetProxies()) > 0
+}
```

Comment:
Could you add a test to future-proof this?  If new fields are added to `apipb.ConfigResponse`, they will be missed by this function.  There's no easy way to alert maintainers to the need to update this function without some kind of automated failure.

It won't be easy.  Maybe some kind of reflection tricks to populate every field and test that this function can catch changes to each?

Reply:
There might be existing solutions for this kind of thing.

Reply:
I can go look at how ` testing/assert` does it.

Reply:
Forgot about `reflect.DeepEqual`..it's actually very easy.

Reply:
> ..it's actually very easy.

Maybe not..I'd need to account for optional fields in `new`..I'll figure it out.

Reply:
It might be easier to have the server return a different status code, like 201, if there are no changes. This way we wouldn't have to update `configIsNew` every time new fields are add or removed. The server already checks if new values needs to be sent back to the client, there's no need for the client to do all the checks again.

Reply:
The server now returns `http.StatusNoContent` if there are no updates to the config.